Create lit test for copyright header validation

AndrewSB opened the pull request:
cc @gottesmm @shahmishal 

I think I have this working when I pass one file into the script, whats the right way to spell "give me all the files so I can `xargs` them into a script" in lit?

AndrewSB committed:
validation-test: Implement date logic for copyright validator

AndrewSB committed:
validation-test/copyright: Implement copyright header testing logic

benrimmington commented:
This could be useful as a script in [swift/utils](https://github.com/apple/swift/tree/master/utils), for updating the copyright notices of all source files.

It could use the `--follow` option to get file history beyond renames, and possibly the `--find-renames=NN%` option to adjust the similarity index.

For example, the commit logs for [Bool.swift](https://github.com/apple/swift/blob/master/stdlib/public/core/Bool.swift) will end with:

* apple/swift@29cebdf797b01b5214f9057934e3f6c33f4221fa (June 2013) if using `--find-renames=50%` (the default)
* apple/swift@249d66f1cdfa143a11c2ce57f93b3cd9c716720d (June 2012) if using `--find-renames=33%`

```bash
git log               \
  --find-renames=33%  \
  --follow            \
  -- "./stdlib/public/core/Bool.swift"
```

I've filed [SR-7507](https://bugs.swift.org/browse/SR-7507) to improve the code templates, so that new source files can use `Copyright (c) 2018` instead of `Copyright (c) 2014 - 2018`.

AndrewSB committed:
validation-test/copyright: Add lit header

purarue committed:
validation-test/copyright: migrated bash script to python

Script that changes copyright headers
 to conform to their git creation/modification dates

futurejones commented:
Can one of the admins verify this patch?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

tbkka commented on validation-test/Python/copyright-conformance.swift:
```diff
@@ -0,0 +1,3 @@
+// Checks to make sure all copyright headers conform with their
+// git creation/last modification date for silly legal reasons.
```
Please remove the text "for silly legal reasons".  Only lawyers are allowed to opine on whether legal issues are silly or not.  😄

AndrewSB replied:
quite sage

committed 👍

tbkka commented:
@swift-ci Please test

AndrewSB committed:
legal reasons are ostensibly not silly

tbkka commented:
@swift-ci Please test

tbkka commented:
Looks like we have a source file somewhere that has malformed UTF-8?  Unfortunately, the script just crashes without reporting the name of the file.  We should fix that.
```
******************** TEST 'Swift-validation(linux-x86_64) :: Python/copyright-conformance.swift' FAILED
: 'RUN: at line 3';   python3 /home/build-user/swift/utils/copyright-conformance.py -rf /home/build-user/build/buildbot_linux/swift-linux-x86_64
--
Exit Code: 1

Command Output (stderr):
--
  File "/home/build-user/swift/utils/copyright-conformance.py", line 84, in has_copyright_notice
    contents = f.read()
  File "/usr/lib/python3.8/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc8 in position 8: invalid continuation byte
```

tbkka commented on utils/copyright-conformance.py:
```diff
@@ -0,0 +1,226 @@
+def has_copyright_notice(candidate_filepath):
+    """Checks if a file has a copyright notice"""
+    with open(candidate_filepath, "r") as f:
+        contents = f.read()
```
This line can fail.  We should catch any error that happens here and print out the `candidate_filepath`

AndrewSB replied:
yup, i see. it looks like some non-unicode files have been added to the tree in the last 4 years

i've pushed a change that skips all files that fail to read with a non-unicode error

AndrewSB committed:
validation-test/copyright: update invocation to use python argument

AndrewSB committed:
utils/copyright-conformance: use a threadpool to speed up git checks

AndrewSB commented:
@tbkka i also went ahead and tested the script. it corrected 8785 files, you can see them here https://github.com/apple/swift/commit/cceb8e51fac401db6ee59e32f72479e3ed81a39e
